btp.lua

```diff
@@ -0,0 +1,289 @@
+        tree:add_le(auth_token, buffer:range(9, 32))
+        tree:add_le(heartbeat_interval, buffer:range(41, 1))
+    elseif mt == "K" then
+        tree:add_le(persist_orders, buffer:range(1, 1))
+    elseif mt == "A" then
+        -- stub
+    elseif mt == "R" then
+        tree:add_le(reject_reason, buffer:range(1, 1))
+    end
+end
+
+function dissect_market_state(buffer, pinfo, tree)
+    tree:add_le(market_state, buffer:range(0, 1))
+    tree:add_le(product_id, buffer:range(1, 8))
+end
+
+function dissect_heartbeat(buffer, pinfo, tree)
+    -- stub
+end
+
+function dissect_disconnect(buffer, pinfo, tree)
+    tree:add_le(disconnect_reason, buffer:range(0, 1))
+    tree:add_le(expected_sequence_id, buffer:range(1, 4))
+    tree:add_le(actual_sequence_id, buffer:range(5, 4))
+end
+
+-- table of body encodings to body dissectors
+body_dissectors = {
+    ["OE"] = dissect_order_entry,
+    ["DC"] = dissect_drop_copy,
+    ["PF"] = dissect_pricefeed,
+    ["LG"] = dissect_login,
+    ["MS"] = dissect_market_state,
+    ["HB"] = dissect_heartbeat,
+    ["DN"] = dissect_disconnect
+}
+
+-- Invoke the body dissector for the given encoding type
+function dissect_body(encoding, buffer, pinfo, tree)
+    body_dissectors[encoding](buffer, pinfo, tree)
+end
+
+-- List of BTP fields
+btp_proto.fields = {
+    -- Header fields
+    protocol_id, version, sequence_id, body_encoding, body_length,
+    -- Disconnect fields
+    disconnect_reason, expected_sequence_id, actual_sequence_id,
+    -- Login fields
+    message_type, connection_id, auth_token, heartbeat_interval, persist_orders,
+    reject_reason
+}
+
+-- This function dissects BTP packets
+function btp_proto.dissector(buffer, pinfo, tree)
+    length = buffer:len()
+    if length < 12 then -- ignore packets that are less than 12 bytes
+        return
+    end
+
+    pinfo.cols.protocol = btp_proto.name
+
+    -- dissect header
+    subtree = tree:add(btp_proto, buffer:range(), "Bitnomial Trading Protocol")
+    header = subtree:add_le(btp_proto, buffer:range(0, 12), "BTP Header")
+    header:add_le(protocol_id, buffer:range(0, 2))
+    header:add_le(version, buffer:range(2, 2))
+    header:add_le(sequence_id, buffer:range(4, 4))
+    header:add_le(body_encoding, buffer:range(8, 2))
+    header:add_le(body_length, buffer:range(10, 2))
+
+    -- dissect body based on the encoding
+    body = subtree:add_le(btp_proto, buffer:range(12), "BTP Body")
+    encoding = buffer:range(8, 2):string()
+    dissect_body(encoding, buffer:range(12), pinfo, body)
+end
+
+-- This function matches packets against a series of tests to determine whether
+-- or not the packet is a BTP packet
+local function heuristic_checker(buffer, pinfo, tree)
+    -- ensure that we have at least the minimum packet size
+    length = buffer:len()
+    if length < 12 then return false end
+
+    -- ensure that the protocol ID matches
+    protocol = buffer:range(0, 2):string()
+    if protocol ~= "BT" then return false end
```

Comment:
Should we check the protocol version too?

Reply:
That sounds right to me, I added the check